Convert dagobah version of the Data Managers tutorial (#3778)
galaxyproject/training-material: topics/admin/tutorials/reference-genomes/tutorial.md

---
layout: tutorial_hands_on

title: "Reference Data with Data Managers"
zenodo_link: ""
questions:
objectives:
  - Have an understanding of the way in which Galaxy stores and uses reference data
  - Be able to download and use data managers to add a reference genome and its pre-calculated indices into the Galaxy reference data system
  - Use an Ansible playbook for all of the above
time_estimation: "1h"
key_points:
  - Understand how Galaxy stores and uses its reference data
  - Understand how to manually add a reference genome and tool indices if required
  - Understand and how to use data managers to make all of this much much easier
contributors:
  - Slugger70
  - afgane

>   - natefoo

Reviewer: If you like can use our fancy new contributions syntax to more accurately record contributions to a training material, declaring which people should be recorded for authorship or just editing.

```
contributions:
  authorship:
    - shiltemann
  editing:
    - bebatut
    - hexylena
  funding:
    - carpentries
  testing:
    - userX
  ux:
    - userY
  infrastructure:
    - userZ
```

see the [full gtn tutorial schema documentation](https://training.galaxyproject.org/training-material/topics/contributing/tutorials/schemas/tutorial.html) for details on the individual roles.

Author: Thanks - I am not sure about the exact contributions. From what I can tell about [the blame](https://github.com/galaxyproject/admin-training/blame/2018-oslo/sessions/05-reference-genomes/ex1-reference-genomes.md) the original was written by Simon, with later additions and updates from Enis.

        - 'reference_source|requested_dbkey': '{{ item }}'
      # Items refere to a list of variables you want to run this data manager. You can use them inside the param field with {{ item }}
      # In case of genome for example you can run this DM with multiple genomes, or you could give multiple URLs.
      items:
        - sacCer3
        #- dm3
        #- mm9
      # Name of the data-tables you want to reload after your DM are finished. This can be important for subsequent data managers
      data_table_reload:
        - all_fasta
        - __dbkeys__
```

Install the fetch DM:

```console
$ shed-tools install -g https://gat-0.student.galaxy.training --name data_manager_fetch_genome_dbkeys_all_fasta --owner devteam -a abbacadabba
Storing log file in: /tmp/ephemeris_2qpg_hrq
(1/1) repository data_manager_fetch_genome_dbkeys_all_fasta already installed at revision 4d3eff1bc421. Skipping.
Installed repositories (0): []
Skipped repositories (1): [('data_manager_fetch_genome_dbkeys_all_fasta', '4d3eff1bc421')]
Errored repositories (0): []
All repositories have been installed.
Total run time: 0:00:00.770248
```

Run the fetch DM:

> ```console
> $ run-data-managers --config fetch-sacCer3.yml -g https://gat-0.student.galaxy.training -a abbacadabba
> Storing log file in: /tmp/ephemeris_kfsmjk2a
> Running data managers that populate the following source data tables: ['all_fasta']
> Dispatched job 1. Running DM: "toolshed.g2.bx.psu.edu/repos/devteam/data_manager_fetch_genome_dbkeys_all_fasta/data_manager_fetch_genome_all_fasta_dbkey/0.0.4" with parameters: {'dbkey_source|dbkey': 'sacCer3', 'reference_source|reference_source_selector': 'ucsc', 'reference_source|requested_dbkey': 'sacCer3'}
> Job 1 finished with state ok.
> Running data managers that index sequences.
> Finished running data managers. Results:
> Successful jobs: 1
> Skipped jobs: 0
> Failed jobs: 0
> ```

Reviewer: ```suggestion
> <code-in-title></code-in-title>
> ```
> run-data-managers --config fetch-sacCer3.yml -g https://gat-0.student.galaxy.training -a abbacadabba
> ```
{: .code-in}

> <code-out-title></code-out-title>
> ```
> Storing log file in: /tmp/ephemeris_kfsmjk2a
> Running data managers that populate the following source data tables: ['all_fasta']
> Dispatched job 1. Running DM: "toolshed.g2.bx.psu.edu/repos/devteam/data_manager_fetch_genome_dbkeys_all_fasta/data_manager_fetch_genome_all_fasta_dbkey/0.0.4" with parameters: {'dbkey_source|dbkey': 'sacCer3', 'reference_source|reference_source_selector': 'ucsc', 'reference_source|requested_dbkey': 'sacCer3'}
> Job 1 finished with state ok.
> Running data managers that index sequences.
> Finished running data managers. Results:
> Successful jobs: 1
> Skipped jobs: 0
> Failed jobs: 0
> ```
```

---

# Overview

**The problem**

The Galaxy server administrator needed to know how to update each type of reference data, how to run the indexers, where to get the data from, and how to update Galaxy's *Tool Data Table* and *location* configuration files.

**Data managers to the rescue**

Data Managers are a special class of Galaxy tool which allows for the download and/or creation of data that is stored within Tool Data Tables and their underlying flat location (e.g. .loc) files. These tools handle the creation of indices and the addition of entries/lines to the data table / .loc file via the Galaxy admin interface.

Data Managers can be defined locally in the data manager and tool data table configuration files or installed through the Tool Shed. When Data Managers are installed from the Tool Shed, their configuration is added to the shed versions of the data manager configuration and tool data table configuration files.

They are a flexible framework for adding reference data to Galaxy (not just genomic data). They are workflow compatible and can run via the Galaxy API. Data managers automatically update the appropriate location files when new data are installed.

> More detailed background information on data managers can be found [here](https://galaxyproject.org/admin/tools/data-managers/), and details on how to define a data manager for a tool can be found [here](https://galaxyproject.org/admin/tools/data-managers/how-to/define/). There is also discussion of this topic in [the Reference Genomes in Galaxy slides]({% link topics/admin/tutorials/reference-genomes/slides.html %}).

Reviewer: ```suggestion
For more background details on data managers, check out:
- [The data managers documentation on the hub](https://galaxyproject.org/admin/tools/data-managers/)
- [How to define a data manager for a tool](https://galaxyproject.org/admin/tools/data-managers/how-to/define/)
- There is also discussion of this topic in [the Reference Genomes in Galaxy slides]({% link topics/admin/tutorials/reference-genomes/slides.html %})
```

>
> If your data are not available as part of the CVMFS repository, Galaxy Data Managers can be used to locally install and build reference data.
>
{: .comment}

> <agenda-title></agenda-title>
>
> 1. TOC
> {:toc}
>
{: .agenda}

# Galaxy Reference Data Components

As discussed in the overview, Galaxy Data Managers handle the population of numerous Galaxy components. From the highest level to the lowest:

- Addition of tool data table entries into the shed version of the tool data table configuration file, `shed_tool_data_table_conf.xml`. Multiple data tables with the same `name` attribute are automatically merged by Galaxy.
- Population of the corresponding location (`.loc`) file when new reference data are installed. The `.loc` file used corresponds to the Data Manager and its specific version that was executed to install the data. The path to the location file can be found in the Data Manager's entry in the tool data table configuration file (above).
- Generation of the reference data on disk, underneath the directory specified by `tool_data_path` in `galaxy.yml`.

# Installing and Using Data Managers with the Galaxy UI

> ## Install a the Fetch Genome Data Manager from the Tool Shed

Reviewer: These always make me deeply uncomfortable for some strange reason and I write some useless blah blah to go here :sweat_smile: 

Good to not bother, it's unnecessary.

Author: You mean text in between headers?

Reviewer: yeah just an aside.

Author: I don't disagree, haha

>
> No special configuration for Galaxy beyond creating an admin user is necessary in order to use Data Managers. However, you may want to be aware of the `enable_data_manager_user_view` option, which enables unprivileged user to browse data installed via Data Managers. This option is not required, and users will be able to *use* data installed via Data Managers regardless.
>
> Data are installed in the path specified by `tool_data_path` in `galaxy.yml`. If you prefer to keep hand-managed reference data separate from DM-managed reference data, you can set `galaxy_data_manager_data_path` instead.
>
{: .comment}

> <tip-title>Intall with Ephemeris</tip-title>
> This hands-on exercise installs data managers and reference data through the Galaxy UI, but you are encouraged to install tools in a deterministic, recordable way through the use of Ephemeris, which is described in [the Installing and Using Data Managers from the command line with Ephemeris section](#installing-and-using-data-managers-from-the-command-line-with-ephemeris) below.
{: .tip}

We will install a data manager that can fetch the various genome sequences from multiple sources.

> > <hands-on-title>Install the Fetch Genome Data Manager </hands-on-title>

Reviewer: ```suggestion
> <hands-on-title>Install the Fetch Genome Data Manager</hands-on-title>
```

> > >             <tool_shed>toolshed.g2.bx.psu.edu</tool_shed>
> > >             <repository_name>data_manager_fetch_genome_dbkeys_all_fasta</repository_name>
> > >             <repository_owner>devteam</repository_owner>
> > >             <installed_changeset_revision>4d3eff1bc421</installed_changeset_revision>
> > >         </tool_shed_repository>
> > >     </table>
> > >     <table name="__dbkeys__" comment_char="#">
> > >         <columns>value, name, len_path</columns>
> > >         <file path="/srv/galaxy/var/tool-data/toolshed.g2.bx.psu.edu/repos/devteam/data_manager_fetch_genome_dbkeys_all_fasta/4d3eff1bc421/dbkeys.loc"/>
> > >         <tool_shed_repository>
> > >             <tool_shed>toolshed.g2.bx.psu.edu</tool_shed>
> > >             <repository_name>data_manager_fetch_genome_dbkeys_all_fasta</repository_name>
> > >             <repository_owner>devteam</repository_owner>
> > >             <installed_changeset_revision>4d3eff1bc421</installed_changeset_revision>
> > >         </tool_shed_repository>
> > >     </table>
> > > </tables>
> > > ```
> > {: .code-out.code-max-300}
> >
> {: .solution }
>
{: .question}

## Download and install a reference genome sequence

> Next, we will install some reference data. Specifically, we will grab sacCer2 (version 2 of the Saccharomyces cerevisiae genome).

Reviewer: ```suggestion
Next, we will install some reference data. Specifically, we will grab sacCer2 (version 2 of the *Saccharomyces cerevisiae* genome).
```

> <hands-on-title>Download and install sacCer2</hands-on-title>
>
> 1. Access the Admin menu from the top bar
> 2. Click **Local Data**, which can be found on the left, under **Server**
>
>    You should see something like this:
>
>    ![nearly empty data manager tool list](../../images/dm-admin-local-data.png)
>
>    If you instead see the message "You do not currently have any Data Managers installed," you may need to restart your Galaxy server.
>
> 3. Click **all_fasta** under **View Tool Data Table Entries**
>
>    You should see the current contents of `tool-data/all_fasta.loc`, which will be empty.
>
> 4. Click **Local Data** to return to the previous page
>

> > 5. Click {% tool [Create DBKey and Reference Genome - fetching]({{data_manager_fetch_genome_all_fasta_dbkey}}) %} under **Installed Data Managers**.

Reviewer: ```suggestion
> 5. Click {% tool [Create DBKey and Reference Genome - fetching](data_manager_fetch_genome_all_fasta_dbkey) %} under **Installed Data Managers**.
```

as far as I can tell this isn't a variable? If it is a variable there's a "NOQA" type statement we can add to this tutorial.

Author: Definitely not a variable, this was me probably miscopying the syntax for tool links from another tutorial.

>
>    > <details-title>What do these fields mean?</details-title>
>    >
>    > The data manager tool form fields correspond to the columns in the tool data table, as referenced in `shed_tool_data_table_conf.xml`.
>    >
>    > Tool Field                | Tool Data Table Entry         | Purpose
>    > ---                       | -----                         | ---
>    > *ID for sequence*         | `value` column in `all_fasta` | Identifier of specific build of the genome, if it differs from the DBKEY. Used for variant builds such as the `hg19female` build of `hg19`.
>    > *DBKEY to assign to data* | `dbkey` column in `all_fasta` | Identifier of the genome, the UCSC build ID for UCSC builds like `hg19`.
>    > *Name of sequence*        | `name` column in `all_fasta`  | Display name of the genome, displayed in Galaxy tool dropdowns.
>    {: .details}
>
> 5. Click **Execute**. In your history, you will see a new dataset for the data manager run. When the job has finished, go back to the Data Manager view on the Galaxy Admin page (Click **Local Data**).
> 6. Click **all_fasta** under *View Tool Data Table Entries*
>
>    You should see that *sacCer2* has been added to all_fasta.
>

> >    ![populated all_fasta data table](../../images/dm-all-fasta-populated.png)

Reviewer: ```suggestion
>    ![populated all_fasta data table which now includes sacCer2 in the genome list](../../images/dm-all-fasta-populated.png)
```

In this part we will repeat the same process as when we installed the Fetch Genome data manager, except that we will install the BWA/BWA-MEM data manager this time.

> <hands-on-title>Install the BWA/BWA-MEM Data Manager </hands-on-title>
>
> 1. Access the **Admin** menu from the top bar
> 2. Click **Install and Uninstall**, which can be found on the left, under **Tool Management**
> 3. Enter `bwa_mem_index` in the search interface
> 4. Click on the first hit, having `devteam` as owner
> 5. Click the **Install** button for the latest revision
{: .hands_on}

## Build the BWA index for a reference genome

In this part we will actually build the BWA index for sacCer2. It will automatically be added to our list of available reference genomes in the BWA tool.

> <hands-on-title>Build the sacCer2 BWA index</hands-on-title>
>
> 1. Access the Admin menu from the top bar
> 2. Click **Local Data**, which can be found on the left, under **Server**

> > 3. Click **{% tool [BWA-MEM index - builder]({{data_manager_bwa_mem_index_builder}}) %}** under **Installed Data Managers**.

Reviewer: ```suggestion
> 3. Click **{% tool [BWA-MEM index - builder](data_manager_bwa_mem_index_builder) %}** under **Installed Data Managers**.
```

>    > param `dbkey_source|dbkey`                         | Value of `<param name="dbkey" ...>` in `<conditional name="dbkey_source">`.
>    > param `sequence_name`                              | Value of `<param name="sequence_name" ...>`.
>    > param `reference_source|reference_source_selector` | Value of `<param name="reference_source_selector" ...>` in `<conditional name="reference_source">`.
>    > param `reference_source|requested_dbkey`           | Value of `<param name="requested_dbkey" ...>` in `<conditional name="reference_source">`.
>    > `items`                                            | A list of variables to template in to `params`, referenced in param fields with {% raw %}`{{ item }}`{% endraw %}. In the case of genomes, for example, you can run this DM with multiple genomes, or you could give multiple URLs.
>    > `data_table_reload`                                | Names of the data tables you want to reload after your DMs are finished running. This can be important for subsequent data managers.
>    >
>    {: .details}
>
>
> 2. Run the Genome Fetch DM with `run-data-managers`:
>
>    > <code-in-title>input: bash</code-in-title>
>    > ```console

> >    > run-data-managers -g https://your-galaxy -a <api-key> --config fetch-sacCer3.yml

Author: I approve! But FWIW I did this because it's used in a bunch of other tutorials. I can make a separate PR updating those.

But also I prefer `example.org` over `example.com`, any objection?

> >    > run-data-managers -g https://your-galaxy -a <api-key> --config fetch-sacCer3.yml

Reviewer: >  I can make a separate PR updating those.

just toss it in this one it's fiiiineee

> any objection?

none
